rl4co/envs/scheduling/fjsp/env.py

```diff
@@ -0,0 +1,436 @@
+        )
+
+        # predecessor and successor adjacency matrices
+        pred = torch.diag_embed(torch.ones(n_ops_max - 1), offset=-1)[None].expand(
+            *batch_size, -1, -1
+        )
+        # the start of the sequence (of each job) does not have a predecessor, therefore we can
+        # mask all first ops of a job in the predecessor matrix
+        pred = pred * ops_seq_order.gt(0).unsqueeze(-1).expand_as(pred).to(pred)
+        succ = torch.diag_embed(torch.ones(n_ops_max - 1), offset=1)[None].expand(
+            *batch_size, -1, -1
+        )
+        # apply the same logic as above to mask the last op of a job, which does not have a successor. The last job of a job
+        # always comes before the 1st op of the next job, therefore performing a left shift of the ops seq tensor here
+        succ = succ * torch.cat(
+            (ops_seq_order[:, 1:], ops_seq_order.new_full((*batch_size, 1), 0)), dim=1
+        ).gt(0).to(succ).unsqueeze(-1).expand_as(succ)
+
+        # adjacency matrix = predecessors, successors and self loops
+        # (bs, max_ops, max_ops, 2)
+        ops_adj = torch.stack((pred, succ), dim=3)
+
+        td = td.update(
+            {
+                "ops_adj": ops_adj,
+                "job_ops_adj": ops_job_bin_map,
+                "ops_job_map": ops_job_map,
+                # "op_spatial_enc": ops_spatial_enc,
+                "ops_sequence_order": ops_seq_order,
+            }
+        )
+
+        return td, n_ops_max
+
+    def _reset(self, td: TensorDict = None, batch_size=None) -> TensorDict:
+        device = td.device
+        self.to(device)
+
+        td_reset = td.clone()
+
+        td_reset, n_ops_max = self._decode_graph_structure(td_reset)
+
+        # schedule
+        start_op_per_job = td_reset["start_op_per_job"]
+        start_times = torch.zeros((*batch_size, n_ops_max))
+        finish_times = torch.full((*batch_size, n_ops_max), INIT_FINISH)
+        ma_assignment = torch.zeros((*batch_size, self.num_mas, n_ops_max))
+
+        # reset feature space
+        busy_until = torch.zeros((*batch_size, self.num_mas))
+        # (bs, ma, ops)
+        ops_ma_adj = (td_reset["proc_times"] > 0).to(torch.float32)
+        # (bs, ops)
+        num_eligible = torch.sum(ops_ma_adj, dim=1)
+
+        td_reset = td_reset.update(
+            {
+                "start_times": start_times,
+                "finish_times": finish_times,
+                "ma_assignment": ma_assignment,
+                "busy_until": busy_until,
+                "num_eligible": num_eligible,
+                "next_op": start_op_per_job.clone().to(torch.int64),
+                "ops_ma_adj": ops_ma_adj,
+                "op_scheduled": torch.full((*batch_size, n_ops_max), False),
+                "job_in_process": torch.full((*batch_size, self.num_jobs), False),
+                "reward": torch.zeros((*batch_size,), dtype=torch.float32),
+                "time": torch.zeros((*batch_size,)),
```

Comment:
Speaking about this initialize size for single dimension values: `[batch_size, 1]` or `[batch_size]` which one do you think it's better? 🤔

Reply:
puh that's a tough one :D I always decide case by case, trying to minimize the number of required reshaping operations